packages/benchmarks/lit-html/repeat/index.ts

```diff
@@ -0,0 +1,230 @@
+  swapMany: {swapCount: 10, from: 100, to: 800},
+  moveForwardOne: {moveCount: 1, from: 100, to: 800, loopCount: 100},
+  moveForwardMany: {moveCount: 10, from: 100, to: 800},
+  moveBackwardOne: {moveCount: 1, from: 800, to: 100, loopCount: 100},
+  moveBackwardMany: {moveCount: 10, from: 800, to: 100},
+  shuffle: {shuffle: true, loopCount: 5},
+};
+
+// If query params are provided, put that opreation in a `custom` step
+const routine = (Object.keys(params).length === 0) ? preset : {
+  render: {...params}, 
+  custom: {...params}
+};
+
+let gid = 0;
+const container = document.getElementById('container')!;
+type Item = {
+  id: number,
+  text: string
+};
+const createItems = (count: number): Item[] => 
+  new Array(count).fill(0).map(() => ({id: gid, text: `item ${gid++}`}));
+
+const itemTemplates = {
+  li: (item: Item) => html`<li>${item.text}</li>`,
+  input: (item: Item) => html`<div>${item.text}: <input value="${item.text}"></div>`
+}
+
+const methods = {
+  repeat(items: Item[], renderItem: (item: Item) => any) {
+    return html`<ul>${repeat(items, item => item.id, renderItem)}</ul>`;
+  },
+  map(items: Item[], renderItem: (item: Item) => any) {
+    return html`<ul>${items.map(renderItem)}</ul>`;
+  }
+};
+
+
+
+let items: Item[] = [];
+
+for (const [step, values] of Object.entries(routine)) {
+
+  const s = {...defaults, ...values};
+  const renderTemplate = methods[s.method];
+  const renderItem = itemTemplates[s.itemType];
+
+  if (step === 'render') {
+    items = createItems(s.initialCount);
+    performance.mark('render-start');
+    render(renderTemplate(items, renderItem), container);
+    performance.measure('render', 'render-start');
+    performance.mark('update-start');
+  } else {
+    
+    // TODO(kschaaf): Accumulate measurements around just the render
+    // once/if https://github.com/Polymer/tachometer/issues/196 is resolved
+    performance.mark(`${step}-start`);
+
+    for (let i=0; i<s.loopCount; i++) {
+
+      if (s.replace) {
+        items = createItems(items.length);
+      }
+
+      if (s.removeCount) {
+        items = items.filter((_, i) => i < s.from || i >= s.from + s.removeCount);
+        if (s.mirror) {
+          items = items.filter((_, i) => i < items.length - s.from-s.removeCount || i >= items.length - s.from);
+        }
+      }
+
+      if (s.addCount) {
+        items = [...items.slice(0, s.to), ...createItems(s.addCount), ...items.slice(s.to)];
+        if (s.mirror) {
+          items = [...items.slice(0, -s.to), ...createItems(s.addCount), ...items.slice(-s.to)];
+        }
+      }
+
+      if (s.swapCount) {
+        items = [
+          ...items.slice(0, s.from),
+          ...items.slice(s.to, s.to + s.swapCount),
+          ...items.slice(s.from + s.swapCount, s.to),
+          ...items.slice(s.from, s.from + s.swapCount),
+          ...items.slice(s.to + s.swapCount)
+        ];
+      }
+
+      if (s.moveCount) {
+        if (s.from < s.to) {
```

Comment:
min/max from/to to save duplication?

Reply:
Not sure how that would help? You need to do the slicing in a different order when moving forward vs. backward.  Unless I'm missing a clever trick you're pointing out, might need to see code.

Reply:
Meh, it's readable as is.